### PR TITLE
Preserve complex types in compression

### PR DESCRIPTION
**Pull request by imranq2**
1. Pre-commit Configuration Update
   - The `.pre-commit-config.yaml` has been significantly expanded with more comprehensive linting, security, and code quality checks.
   - Added hooks for:
     * Basic file checks (end-of-file, trailing whitespace)
     * Conflict detection
     * Security scans (AWS credentials, private keys)
     * Branch protection
     * Dependency and import checks
   - Replaced `autoflake` and `black` with `ruff` for linting and formatting
   - Added `mypy` with stricter type checking
   - Added `bandit` for security scanning
   - Added `detect-secrets` for secret detection

2. Dependency Updates
   - Added `bandit` for security checks
   - Added `ruff` for linting and formatting
   - Updated various package versions

3. Code Changes
   - Multiple files have been refactored with minor improvements:
     * Simplified string representations
     * Added type hints
     * Improved error handling
     * Introduced new utility classes for type preservation and string compression

4. New Utility Classes
   - Added `TypePreservationSerializer`, `TypePreservationEncoder`, and `TypePreservationDecoder` for advanced JSON serialization
   - Added `StringCompressor` for efficient string compression
   - Added `OrderedDictToDictConverter` for converting OrderedDict to regular dict

5. Test Coverage
   - Comprehensive test suites added for new utility classes
   - Tests cover various scenarios including edge cases, performance, and error handling

Key Improvements:
- Enhanced code quality checks
- More robust type handling
- Better serialization and compression utilities
- Improved security scanning

### CODE REVIEW

File: compressedfhir/fhir/fhir_resource.py [hunk 1]
```diff
@@ -161,3 +134,40 @@ def meta(self, value: FhirMeta | None) -> None:
         else:
             assert isinstance(value, FhirMeta)
             self["meta"] = value.dict()
+
+    @classmethod
+    @override
```
[the remark this answers is not in the document]
Comment (imranq2) [reply compressedfhir/fhir/fhir_resource.py:112]: override is imported in this file

**Review by alvinhenrick — APPROVED** [review]
Looks good. 

One suggestion may be we could just use `jsonpickle` 

1. Built-in support for complex type preservation
2. Handles nested objects
3. Less custom code
4. Widely used in Python ecosystem

Something like this :

```
import jsonpickle

class TypePreservationSerializer:
    @staticmethod
    def serialize(obj):
        return jsonpickle.encode(obj, unpicklable=True)
    
    @staticmethod
    def deserialize(json_str):
        return jsonpickle.decode(json_str)
        ```
